vpr/src/route/rr_graph_reader.cpp

```diff
@@ -866,75 +724,467 @@ void process_rr_node_indices(const DeviceGrid& grid) {
- * to their unique cost index identifier. CHANX and CHANY cost indicies are set after the
- * seg_id is read in from the rr graph*/
-void set_cost_indices(pugi::xml_node parent, const pugiutil::loc_data& loc_data, const bool is_global_graph, const int num_seg_types) {
+/**
+ * Load a Cap'n Proto rr_graph.
+ */
+void load_rr_file_capnp(const t_graph_type graph_type,
+                        const DeviceGrid& grid,
+                        const std::vector<t_segment_inf>& segment_inf,
+                        const enum e_base_cost_type base_cost_type,
+                        int* wire_to_rr_ipin_switch,
+                        const char* read_rr_graph_name) {
     auto& device_ctx = g_vpr_ctx.mutable_device();
 
-    //set the cost index in order to load the segment information, rr nodes should be set already
-    for (size_t inode = 0; inode < device_ctx.rr_nodes.size(); inode++) {
-        if (device_ctx.rr_nodes[inode].type() == SOURCE) {
-            device_ctx.rr_nodes[inode].set_cost_index(SOURCE_COST_INDEX);
-        } else if (device_ctx.rr_nodes[inode].type() == SINK) {
-            device_ctx.rr_nodes[inode].set_cost_index(SINK_COST_INDEX);
-        } else if (device_ctx.rr_nodes[inode].type() == IPIN) {
-            device_ctx.rr_nodes[inode].set_cost_index(IPIN_COST_INDEX);
-        } else if (device_ctx.rr_nodes[inode].type() == OPIN) {
-            device_ctx.rr_nodes[inode].set_cost_index(OPIN_COST_INDEX);
+    vtr::ScopedStartFinishTimer timer("Loading routing resource graph");
+    MmapFile f(read_rr_graph_name);
+    /* Increase reader limit to 1G words. */
+    ::capnp::ReaderOptions opts = ::capnp::ReaderOptions();
+    opts.traversalLimitInWords = 1024 * 1024 * 1024;
+    ::capnp::FlatArrayMessageReader reader(f.getData(), opts);
+    auto rr_graph = reader.getRoot<ucap::RrGraph>();
+
+    const char* tool_version = rr_graph.getToolVersion().cStr();
+    if (tool_version && strcmp(tool_version, vtr::VERSION) != 0) {
+        VTR_LOG("\n");
+        VTR_LOG_WARN("This architecture version is for VPR %s while your current VPR version is %s compatability issues may arise\n", vtr::VERSION, tool_version);
+        VTR_LOG("\n");
+    }
+    const char* tool_comment = rr_graph.getToolComment().cStr();
+    std::string correct_string = "Generated from arch file ";
+    correct_string += get_arch_file_name();
+    if (tool_comment && strcmp(tool_comment, correct_string.c_str()) != 0) {
+        VTR_LOG("\n");
+        VTR_LOG_WARN("This architecture version is for VPR %s while your current VPR version is %s compatability issues may arise\n", vtr::VERSION, tool_version);
+        VTR_LOG("\n");
+    }
+
+    /* Decode the graph_type */
+    bool is_global_graph = (GRAPH_GLOBAL == graph_type ? true : false);
+
+    /* Compare the grid with the architecture file to ensure consistency. */
+    auto grid_locs = rr_graph.getGrid().getGridLocs();
+    for (auto grid_loc : grid_locs) {
```

Comment:
The XML and capnproto copy/paste variants with a lot of common code.  Can you think of a way to make the code more common?

Reply:
Example idea: Should there be an API for accessing rr graph data that is then implemented by the XML and capnproto generators, and then is used once, rather than touching each interface?

This need not be done with virtual interfaces and indirection if you make the accessors some form of a generic interface.

Reply:
I can do things with the XML generator, but the capnproto generator only generates the capnp schema, so the capnp interface is pretty much fixed.

I thought that could be fixed by the rr_graph object but it looks like it doesn't have a virtual accessor feature.

An option is changing the xml generator to populate/write out a capnp object instead of generating its own structs. This way we also get an xml<->capnp converter for free. Of course this adds another set of assumptions to the generator, about the API exposed by the capnp schema compiler.

Reply:
@duck2 How goes this change?  I think in the next week or two we are going to want to integrate a variant of your work, and I'd like to incorporate the version that includes this refactoring.

Reply:
I'm currently hand-writing such an interface for the rr_graph to see how would it go. It looks like this:

```c++
template<class T>
void write_rr_graph(T &in, std::ostream &os){
[...]
	os << "<switches";
	os << ">";
	n0 = in.get_switches_num_switches();
	for(i0=0; i0<n0; i0++){
		in.get_switches_grab_switch(i0);
		os << "<switch";
		os << " id=\"" << in.get_switches_switch_id() << "\"";
		os << " name=\"" << in.get_switches_switch_name() << "\"";
		if(in.get_switches_switch_type())
			os << " type=\"" << lookup_switch_type[(int)in.get_switches_switch_type()] << "\"";
		os << ">";
		if(in.has_switches_switch_timing()){
			os << "<timing";
			os << " Cin=\"" << in.get_switches_switch_timing_Cin() << "\"";
			os << " Cinternal=\"" << in.get_switches_switch_timing_Cinternal() << "\"";
			os << " Cout=\"" << in.get_switches_switch_timing_Cout() << "\"";
			os << " R=\"" << in.get_switches_switch_timing_R() << "\"";
			os << " Tdel=\"" << in.get_switches_switch_timing_Tdel() << "\"";
			os << "/>";
		}
		os << "<sizing";
		os << " buf_size=\"" << in.get_switches_switch_sizing_buf_size() << "\"";
		os << " mux_trans_size=\"" << in.get_switches_switch_sizing_mux_trans_size() << "\"";
		os << "/>";
		os << "</switch>";
	}
[...]
}
```

The main problem is that templates are implicit interfaces. (I think virtual fn overhead would start to count at hundreds of millions of calls.) We can provide a "template stub" to the user, but I have no idea about how to get C++ to _enforce_ this interface.

Another problem of the implicit interface is: we can't say that "this template fn should return a pointer to this template, which should have implemented these things". This results in very long function names.

Reply:
> Another problem of the implicit interface is: we can't say that "this template fn should return a pointer to this template, which should have implemented these things". This results in very long function names.

I recommend we define a pure virtual interface that gets inherited from, but then use static_assert and [is_base_of](https://en.cppreference.com/w/cpp/types/is_base_of) to enforce that the template is derived from the virtual interface, but we use the concrete type to enable function inlining and to avoid the v-table lookup.

Reply:
```
n0 = in.get_switches_num_switches();
	for(i0=0; i0<n0; i0++){
```

Probably better to make a range iterator, which there is support for in vtr in libvtrutil.

Reply:
That's a good idea! I wrote it for a simpler schema and it works. Now the problem is with the pointer and list types.

If we take the XSD schema literally, the rr_graph has a "switches" struct which includes only a list of "switch" structs. Then, in order to use this interface, which is very similar to Cap'n Proto's one:
```c++
template<class T>
void write_rr_graph(T &in, std::ostream &os){
[...]
	os << "<switches";
	os << ">";
        auto switches_parent = in.get_switches();
        auto switches = switches_parent.get_switches();
	for(auto& switch_: switches){
		os << "<switch";
		os << " id=\"" << switch_.get_id() << "\"";
		os << " name=\"" << switch_.get_name() << "\"";
		os << ">";
		[...]
	}
[...]
}
```

something like this is needed? Perhaps use `std::begin` and `std::end` instead?

```c++
class SwitchBase {
public:
	virtual inline int get_id(void) = 0;
	virtual inline const char * get_name(void) = 0;
	virtual inline void set_id(int id) = 0;
	virtual inline void set_name(const char *name) = 0;
};

template<class T0>
class SwitchPtrBase {
	static_assert(std::is_base_of<SwitchBase, T0>::value);
	virtual inline T0* begin(void) = 0;
	virtual inline T0* end(void) = 0;
        virtual inline T0 operator [](size_t i) = 0;
        virtual inline size_t size(void) = 0;
        virtual inline void add_switch(T0 &switch_) = 0;
};

template<class T0>
class SwitchesBase {
	static_assert(std::is_base_of<SwitchPtrBase, T0>::value);
	virtual inline T0 get_switches(void) = 0;
        virtual inline void set_switches(T0& switches) = 0;
};

template<class T0>
class RrGraphBase {
public:
	static_assert(std::is_base_of<SwitchesBase, T0>::value);
	virtual inline T0& get_switches(void) = 0;
        virtual inline void set_switches(T0& switches) = 0;
};
```

I can see this easily getting ugly. Maybe long function names and "grab_X" functions are better?

Reply:
> I can see this easily getting ugly. Maybe long function names and "grab_X" functions are better?

For now, let's get something working, and go from there.  If it is easier for you to write the flat interface with long function names, let's start there.

Reply:
Can you push what you have right now?

Reply:
Nothing much but I'll push it to a new branch in uxsdcxx today.

Reply:
I now have a grasp on the load interface. [Here](https://gist.github.com/duck2/758b6de52187def4d82c1d376e3a6de2) is the generated `load_X` functions for the rr_graph(though nothing compiles right now). See https://github.com/duck2/uxsdcxx/tree/interface-consumer for what I have so far.

Reply:
So far, the interface is like this:
`inline void add_rr_nodes_node(void)`: Expects the user to add a node and keep internal state on the most recently added node.
`inline void init_node_loc(void)`: Useless for VPR but might be needed if one needs to allocate memory for child structs.
`inline void set_node_loc_ptc(int)`: Should be straightforward if the user kept state about most recent node.

Reply:
We definitely want `add_*` functions to return void *, and have that passed back to the init_node_loc / set_node_loc_ptr.

Reply:
Right. That's a good way to share state between functions.

Reply:
Another thing to add is a resize call.  If the reader knows the number of items, preallocation can be a big win in some cases.

Reply:
I can now generate a base class for loading xml files, which also gets inlined when compiled with `-O3`. Tested it only for simple schemas, though.

The generated code for rr_graph at this stage can be found [here](https://gist.github.com/duck2/fd5b99c866d2e4779667635e5b844c10). It goes through g++ without any trouble right now.

Currently all code is in the header since the template fn implementation has to be seen from the caller's compilation unit. That brings everything else in.

Reply:
I've tried to integrate the existing interface and found a couple changes I think would make things work better:

 - I think attributes on an tag opening should be passed as one (either as arguments or as a struct) into the interface.  In particular the `id="#"` attribute really needs to be available as soon as possible to avoid gymnastics
 - I think there needs to be a place to `free` void *'s when they are discarded.   Right not there is a an `init_....`, there should probably be a `finish_....`

Reply:
Providing `finish_*` is a good idea. I initially thought the `void *`s would be pointers to existing objects or maybe even indices, not data allocated for this purpose.

So the `init_*` and `add_*` functions should have the XML attributes as arguments? This makes sense, it would significantly reduce the number of functions. What to do in the case of non-existing optional arguments? Pass 0 or `default_value` if there is one?

Reply:
I think only add attribute arguments for required attributes.  So the call order becomes:
`init_*`/`add_*` with all required attributes
`set_*` for any present optional attributes
Deal with children nodes
Call `finish_*`

Reply:
FYI, the reason for `finish_*` is that some of these calls are not setting data, but instead verifying data.  See `verify_blocks`, etc.